### DISCUSSION TITLE
Question about the performance of isio and envoy

### DISCUSSION THREAD

**Discussion by eottabom** [istio/istio#51333] [Q&A] [answered]
Can I ask you a question here?

I have a question about envoy in the istio.

istio is using envoy as a proxy,
As far as I know, envoy is in c++, and istio is using envoy by wrapping it in go.

Is this right?

If this is right,
Is there a difference in performance between using istio and using envoy in the service mesh structure?

**Comment by howardjohn** [comment, marked as answer]
Istio has a few Go binaries that run alongside Envoy, but its just running envoy like normal basically. Technically, we have a small fork of Envoy in https://github.com/istio/proxy/ to add a few extensions.

So they should perform the same, because they are the same. However, you can configure envoy in many many ways -- and those have performance impacts. So you may say "The way Istio configures Envoy is slower than the way <something> configures envoy", but probably not "Istio performance differs from envoy"

**Reply by eottabom** [reply]
Then, let's say it's a basic configuration
The configuration of the envoy service mesh and
Does it mean that there is no difference when configuring the istio service mesh?
I thought envoy in the istio was go lang binary, but I guess it's not.

If the envoy configuration in the istio mesh increases,
I think there will be an impact on performance due to more iptables rules, is this right?

**Reply by howardjohn** [reply]
Iptables performance issues are _generally_ about O(n) scaling, not inherent slowness. Istio has a fix ~10 rules, so doesn't suffer from the same problem kube-proxy has. Note this isn't entirely true in _extremely high performance_ scenarios (like 100Gb/s, which is orders of magnitude beyond what Envoy will do with HTTP) but for the vast majority iptables is not relevant to Istio performance.

As to the question of Istio vs Envoy -- this is like asking "Is envoy faster than C++". They cannot really be compared unless you have configuration for each. Istio isn't really doing anything that is very expensive on top of how someone would _typically_ configure Envoy, though, with a possible exception of mTLS.

**Reply by eottabom** [reply]
Thank you for your confirmation.
